Packet++/src/SmtpLayer.cpp

```diff
@@ -0,0 +1,228 @@
+
+		return static_cast<SmtpCommand>(val);
+	}
+
+	std::string SmtpRequestLayer::getCommandString() const { return getCommandInternal(); }
+
+	bool SmtpRequestLayer::setCommandOption(const std::string &value) { return setCommandOptionInternal(value); }
+
+	std::string SmtpRequestLayer::getCommandOption(bool removeEscapeCharacters) const
+	{
+		if (removeEscapeCharacters)
+		{
+			std::stringstream ss;
+			std::string field = getCommandOptionInternal();
+			for (size_t idx = 0; idx < field.size(); ++idx)
+			{
+				if (int(field.c_str()[idx]) < 127 && int(field.c_str()[idx]) > 31) // From SPACE to ~
+					ss << field.c_str()[idx];
+			}
+			return ss.str();
+		}
+		return getCommandOptionInternal();
+	}
+
+	std::string SmtpRequestLayer::getCommandInfo(SmtpCommand code)
+	{
+		switch (code)
+		{
+		case SmtpCommand::DATA:
+			return "Starting mail body";
+		case SmtpCommand::EHLO:
+			return "Initiate conversation";
+		case SmtpCommand::EXPN:
+			return "Expand the mailing list";
+		case SmtpCommand::HELO:
+			return "Initiate conversation";
+		case SmtpCommand::HELP:
+			return "Ask information";
+		case SmtpCommand::MAIL:
+			return "Sender indication";
+		case SmtpCommand::NOOP:
+			return "No operation";
+		case SmtpCommand::QUIT:
+			return "Close conversation";
+		case SmtpCommand::RCPT:
+			return "Receiver indication";
+		case SmtpCommand::RSET:
+			return "Abort transaction";
+		case SmtpCommand::VRFY:
+			return "Identify user";
+		case SmtpCommand::STARTTLS:
+			return "Start TLS handshake";
+		case SmtpCommand::TURN:
+			return "Reverse the role of sender and receiver";
+		case SmtpCommand::SEND:
+			return "Send mail to terminal";
+		case SmtpCommand::SOML:
+			return "Send mail to terminal or to mailbox";
+		case SmtpCommand::SAML:
+			return "Send mail to terminal and mailbox";
+		case SmtpCommand::AUTH:
+			return "Authenticate client and server";
+		case SmtpCommand::ATRN:
+			return "Reverse the role of sender and receiver";
+		case SmtpCommand::BDAT:
+			return "Submit mail contents";
+		case SmtpCommand::ETRN:
+			return "Request to start SMTP queue processing";
+		case SmtpCommand::XADR:
+			return "Release status of the channel";
+		case SmtpCommand::XCIR:
+			return "Release status of the circuit checking facility";
+		case SmtpCommand::XSTA:
+			return "Release status of the number of messages in channel queues";
+		case SmtpCommand::XGEN:
+			return "Release status of whether a compiled configuration and character set are in use";
+		default:
+			return "Unknown command";
+		}
+	}
+
+	std::string SmtpRequestLayer::getCommandAsString(SmtpCommand code)
+	{
+		std::stringstream oss;
+		for (size_t idx = 0; idx < 8; ++idx)
+		{
+			char val = (uint64_t(code) >> (8 * idx)) & UINT8_MAX;
+			if (val) // Dont push if it is a null character
+			{
+				oss << val;
+			}
+		}
+		return oss.str();
+	}
+
+	std::string SmtpRequestLayer::toString() const { return "SMTP Request: " + getCommandString(); }
```

Comment:
As I mentioned in a previous comment, `getCommandString()` might return an empty string. We should take it into consideration here

Reply:
I'm not sure I understand your mention about previous comment. Which one?

Reply:
I meant this comment: https://github.com/seladb/PcapPlusPlus/pull/1201#discussion_r1407354460

If the command is empty, `toString()` will be: `"SMTP Request: "` which is weird...

Reply:
If it is not weird enough it currently prints the username or password for these packets instead of empty string😁

![image](https://github.com/seladb/PcapPlusPlus/assets/64282645/0b71ad1d-db3f-499b-a0bf-b37ad820ee06)

I can first check the length of the command string to remove `:` but I should check why I design the algorithm to decide all of the data as command when there is no hyphen or space since I can't remember for now. If I can change to decide data as option it will return empty string so i can remove it🤔